Feature difx2fits freq filtering

jwagnerhki opened the pull request:
Hi Walter, could you have a look at these changes?

They add a command line option to difx2fits,

$ difx2fits --help
  --freqlist <list>
                      Only propagate IFs with listed low edge freqs (comma separated, MHz)

With that, one can trim down the set of frequencies (& visibility records) that are going exported into the FITS-IDI file.

A usage example would be:

$ difx2fits -d -v -v -v --freqlist 86012,86076,86140,86204,86268,86332,86396,86460  example.fits

The above is for a test case in Bonn. DiFX correlation produced 64 MHz wide bands at sixteen sky frequencies 86012, 86076, 86140, 86204, 86268, 86332, 86396, 86460, 86524, 86588, 86652, 86716, 86780, 86844, 86908 MHz (low edges).

In a subset of the VLBI array only the first eight frequencies contain desired visibility data. 

Those data can now be exported in isolation using the new --freqlist option.

The resulting FITS-IDI has a clean frequency table without "errant" IF entries that'd lack visibility data. Such IFs in the frequency table but without data causes major issues in CASA, one project PI reported...

walterfb commented:
Hi Jan,

A few comments:

1. There is a change to updateDifxInput() which is unfortunate, but this
action is quite analogous to the merge options, so it makes sense to do
this.   You should try compiling all of the DiFX tree to see if there are
other programs that need an update to this function call.

2. There are a few capitalizations that I would suggest changing:
"jobfreqIdRemap" should be "jobFreqIdRemap", and "Datafilter" should be
"DataFilter"

3. In the new Datafilter structure there is the line: "char
**includeAntennasList; /* TODO; if not NULL, ignore all antennas not found
in this NULL-terminated list (2-letter codes, case insensitive) */" .  The
antenna names are not necessarily 2-letters long (e.g., VLA is "Y").  More
generally, these should be the names carried by DiFX, which could in theory
be any number of characters long and could include numbers.  I think for
now only a change to the comment is needed.

Otherwise I think this _looks_ good.  I haven't tried compiling or using
it.  It does look to be fully safe when the feature is not used.

Is it possible for you to update the pull request with the minor changes
above?

-Walter

On Fri, Nov 10, 2023 at 6:42 AM Jan Wagner ***@***.***> wrote:

> @jwagnerhki <https://github.com/jwagnerhki> requested your review on: #28
> <https://github.com/difx/difx/pull/28> Feature difx2fits freq filtering.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/difx/difx/pull/28#event-10925325246>, or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAFOKRWNA54ZOEU65NV565LYDYVODAVCNFSM6AAAAAA7GGLH3GVHI2DSMVQWIX3LMV45UABCJFZXG5LFIV3GK3TUJZXXI2LGNFRWC5DJN5XDWMJQHEZDKMZSGUZDINQ>
> .
> You are receiving this because your review was requested.Message ID:
> ***@***.***>
>

jwagnerhki commented:
Hi Walter, thanks for your review. I've implemented the changes in 52ae112, and updated the ChangeLog in 353db5a.

A test build with

```
PKGS="--ipp --withm6support --withmark6meta --withpolconvert --withpython --withdatasim"
$DIFXROOT/git/install-difx --reconf $PKGS
```
went fine. The generated FITS-IDI for test cases ends up looking fine in AIPS - only the requested frequencies are in the FQ table (PRTAB FQ), and visibility data as well contain only the requested frequencies (PRTUV). Filtering appears to work. I need to ask a VLBI project PI for a more detailed check. Meanwhile perhaps okay to merge back into dev?

Btw, in principle, GitHub should see my two new commits to this branch, and should  place them under this pull request, esp. given the GitHub note a bit higher up wrt "Add more commits by pushing to the feature-difx2fits-freq-filtering branch on difx/difx.". I am not seeing those commits listed at all under https://github.com/difx/difx/pull/28/commits , though. Is that a new normal for GitHub...? I hope the two added commits make it in when merging this pull request...
